CorsixTH/Lua/hospital.lua

```diff
@@ -1217,32 +1217,28 @@ function Hospital:createEmergency(emergency)
     end
 
     self.emergency = emergency
-    -- The last room in the list of treatment rooms is considered when checking for availability.
-    -- It works for all original diseases, but if we introduce new multiple room diseases it might break.
-    -- TODO: Make it work for all kinds of lists of treatment rooms.
-    -- TODO: Change to make use of Hospital:checkDiseaseRequirements
-    local no_rooms = #emergency.disease.treatment_rooms
-    local room_name, required_staff, staff_name =
-      self.world:getRoomNameAndRequiredStaffName(emergency.disease.treatment_rooms[no_rooms])
 
-    local staff_available = self:countStaffOfCategory(required_staff) > 0
-    -- Check so that all rooms in the list are available
-    if self:countRoomOfType(emergency.disease.treatment_rooms[no_rooms]) > 0 then
-      room_name = nil
+    -- Check needed rooms and staff.
+    local room_name, staff_name
+    local required = self:checkDiseaseRequirements(emergency.disease.id)
+    if required then
+      if #required.rooms > 0 then room_name = required.rooms[1] end -- Take the first missing room
+
+      for sn, _ in pairs(required.staff) do staff_name = sn; break; end
```

Comment:
When I added the additional fax, I looked at that TODO and checkDiseaseRequirements, and I think you have the same problem and I just worked around it to avoid a refactor of checkDiseaseRequirements.

Ward and Op Theatre, room and staff will mix in the fax.
ie build_and_employ, Ward and Surgeon or Op Theatre and Nurse depending on rooms built and available staff.

Reply:
I see a few ways out, but not sure what is preferred.

Current strings handle
- 'all covered',
- 'missing 1 person of 1 staff type',
- 'missing 1 room type',
- 'missing 1 person of 1 staff type and 1 room type'.

In the general sense, we can have an arbitrary number of missing room types, and need 1 or more employees of 1 or more staff types.

- One way out is to extend the strings to the point that you can have an arbitrary number missing of room types and/or staff types/persons. Likely this explodes in the number of strings, for English it's not too bad on first judgement, but in translations it may become messy.

- Second way out is to drop the staff things if it gets complicated, and generalize with an additional string to something like "fully staffed X, Y, and Z rooms" (where an arbitrary list of rooms is still a separate problem to deal with in translations). Note that at this point re-use of the checkDiseaseRequirements fails, as it doesn't tell about rooms that you need and already have. This needs to be coded separately or the function needs to be extended.

- Third way out is not to generate emergencies that don't fit in the currently available strings. This may however lead to never having an emergency if you build a sufficiently crippled hospital.

Reply:
In the seek room action its handled a bit differently, as the seek there has multiple rooms, whereas here for emergency in the current state of play, is always the last room to visit for treatment. In seek room there is no attempt for multiple rooms or staff in the message, you just '_can_not_cure_' them.

Could probably achieve the same by processing the checkDiseaseRequirements in reverse, and aborting after the first if a flag is set, or adding the room_id as input and checking against only that room. But that all just feels a little hacky. Or modify checkDiseaseRequirements to reer to the room for the staff as well and modify the callers to process that.

Also the text returned is different long_name vs name, ie Psychiatric Room vs Psychiatric so there is a few things to unpick. The method getRoomNameAndRequiredStaffName appears to have been only added for fax processing.